File: python/Q875_koko_eating_bananas.py
#
# @lc app=leetcode id=875 lang=python3
#
# [875] Koko Eating Bananas
#


# @lc code=start
class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        left: int = 1
        right: int = max(piles)
        result: int = right

        while left <= right:
            speed: int = left + ((right - left) // 2)
            time: int = sum((p // speed) + (p % speed > 0) for p in piles)

            if time <= h:
                right = speed - 1
                result = speed
            else:
                left = speed + 1

        return result


# @lc code=end

File: rust/src/leetcode/Q875_koko_eating_bananas.rs
/*
 * @lc app=leetcode id=875 lang=rust
 *
 * [875] Koko Eating Bananas
 */

// @lc code=start
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut left: i32 = 1;
        let mut right: i32 = *piles.iter().max().unwrap();

        while left <= right {
            let mid: i32 = left + (right - left) / 2;
            let time: i64 = piles
                .iter()
                .map(|&p| (p / mid) as i64 + if p % mid > 0 { 1 } else { 0 })
                .sum();

            if time > h as i64 {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        left
    }
}
// @lc code=end
